My Project/clsTranslateWinForms.vb

```diff
@@ -1 +1,637 @@
+        '  when the translation database was generated.
+        '  For example, during testing we found cases for sub-dialog->tab->group->panel->radioButton 
+        '  similar to:
+        '    Database component name: sdgPlots_tbpPlotsOptions_tbpXAxis_ucrXAxis_grpAxisTitle_rdoSpecifyTitle
+        '    Runtime  component name: sdgPlots_tbpPlotsOptions_tbpXAxis_ucrXAxis_grpAxisTitle_ucrPnlAxisTitle_rdoSpecifyTitle
+
+        'split the full component name into 2 parts: parent names & child name
+        '  e.g 'sdgPlots_tbpPlotsOptions_tbpXAxis_ucrXAxis_grpAxisTitle' & '_rdoSpecifyTitle'
+        Dim iTmpIndex As Integer = strComponentName.LastIndexOf("_")
+        Dim strParentNames As String = strComponentName.Substring(0, iTmpIndex)
+        Dim strControlName As String = strComponentName.Substring(iTmpIndex)
+
+        'if the dictionary contains a single key that matches <parents>_<other controls>_<child>,
+        '  then return the component associated with that key
+        Dim lstComponents As List(Of Component) =
+            dctComponents.Where(Function(x) Regex.IsMatch(x.Key, strParentNames & "\w+_\w+" & strControlName)) _
+            .Select(Function(x) x.Value).ToList()
+        If lstComponents.Count = 1 Then
+            clsComponent = lstComponents(0)
+        End If
+
+        Return clsComponent
+    End Function
+
+    '''--------------------------------------------------------------------------------------------
+    ''' <summary>   
+    '''    Returns <paramref name="strText"/> translated into <paramref name="strLanguage"/>. 
+    '''    <para>
+    '''    Translations can be bi-directional (e.g. from English to French or from French to English).
+    '''    If <paramref name="strText"/> is already in the current language, or if no translation 
+    '''    can be found, then returns <paramref name="strText"/>.         
+    '''    </para></summary>
+    '''
+    ''' <param name="strText">          The text to translate. </param>
+    ''' <param name="strLanguage">      The language code to translate to (e.g. 'fr' for French).
+    '''                                 </param>
+    ''' <param name="clsConnection">    An open connection to the SQLite translation database. </param>
+    '''
+    ''' <returns>   <paramref name="strText"/> translated into <paramref name="strLanguage"/>. </returns>
+    '''--------------------------------------------------------------------------------------------
+    Private Shared Function GetDynamicTranslation(strText As String, strLanguage As String, clsConnection As SQLiteConnection) As String
+        If String.IsNullOrEmpty(strText) Then
+            Return ""
+        End If
+
+        Using clsCommand As New SQLiteCommand(clsConnection)
+
+            'in the translation text, convert any single quotes to make them suitable for the SQL command
+            strText = strText.Replace("'", "''")
```

Comment:
@lloyddewit  this type of check could be replaced if we use parameterised query. In future we may get very odd texts that need to be translated, and such texts could contain sql recognised string literals. We can still choose the option of using the raw query as used here and keep expanding the sql checks as we get more 'problematic' translation texts or we use parametrised queries that will do that for us automatically. 

I'm also happy we leave it as it is then do a separate PR that improves on the generated sql strings, or if it saves time I can post an equivalent parameterised query for you.

Reply:
I would love to improve this.
The parameterized queries you showed me last time we spoke seemed a much better solution.
Please could you raise an issue about this (in the TranslateWinForms repo i.e. this repo)?
If you could include an example parameterized query in the issue, then that would be fantastic.

Reply:
I have now outlined this in issue #6